ubuntu-mainline-kernel.sh

```diff
@@ -384,28 +384,18 @@ load_remote_versions () {
         fi
 
         IFS=$'\n'
-        for line in $remote_html_cache; do
-            [[ "$line" =~ "folder" ]] || continue
+        while read -r line; do
             [[ $use_rc -eq 0 ]] && [[ "$line" =~ -rc ]] && continue
-            [[ "$line" =~ v[0-9]+\.[0-9]+(\.[0-9]+)?(-rc[0-9]+)?/ ]] || continue
             [[ -n "$2" ]] && [[ ! "$line" =~ $2 ]] && continue
-
-            line=${line##*href=\"}
-            line=${line%%\/\">*}
-            [[ ! "$line" =~ (v[0-9]+\.[0-9]+)\.[0-9]+ ]] && [[ "$line" =~ (v[0-9]+\.[0-9]+)(-rc[0-9]+)? ]] && line=${BASH_REMATCH[1]}".0"${BASH_REMATCH[2]}
-
             REMOTE_VERSIONS+=("$line")
-        done
+        done < <(grep -oP '^.*?DIR.*?href="\Kv\d+(?:\.\d+){1,2}(?:-rc\d+)?(?=/")' <<<"$remote_html_cache" | perl -pe 's/^v\d++\.\d++(?!\.)/$&.0/; s/-/~/' | sort -V | perl -pe 's/~/-/')
```

Comment:
Can you explain whats happening here?

Also I would strongly prefer not to add perl as a dependency, you'd have to make a really compelling argument why perl is the only/best solution for this and we cant use more standard bash utils (even when perl is often installed oob on distros, I prefer to keep the dependencies to a minimum)

Reply:
Sure, I chose Perl because I thought it was specified by POSIX, and I prefer its powerful regex syntax.

First, it greps the entire HTML response for lines that have `DIR` followed by a link, much like the original does. The `\K` and the look-ahead assertion limit what's captured so it produces a clean list of versions in a single pass. Then perl adds the extra `.0` for two-digit version numbers, and temporarily modifies `-rc` suffixes so that they're placed *before* stable releases by the `sort`. The last part reverts the temporary suffix change.

Reply:
I've removed the `perl` dependency.